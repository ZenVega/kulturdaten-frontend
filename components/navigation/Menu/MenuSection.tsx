import styled from '@emotion/styled';

export const StyledSection = styled.div<{ variant: MenuSectionVariant }>`
  width: 100%;
  background: var(--grey-200);
  font-size: var(--font-size-400);
  line-height: var(--line-height-400);
  font-weight: 700;
  border-radius: 0.75rem;
  /* border: 1px solid var(--grey-400); */
  overflow: hidden;

  ${({ variant }) =>
    variant === MenuSectionVariant.default
      ? 'box-shadow: 0 0 0.75rem -0.125rem rgba(0, 0, 0, 0.25);'
      : ''}
`;

const StyledSectionHead = styled.div<{
  background?: string;
  color?: string;
  uppercase?: boolean;
}>`
  display: flex;
  flex-direction: row;
  padding: 0 0 0 0.75rem;
  box-shadow: 0px 1px 0px var(--grey-400);

  ${({ background }) => (background ? `background: ${background};` : '')}
  ${({ color }) => (color ? `color: ${color};` : '')}
  ${({ uppercase }) => (uppercase ? `text-transform: uppercase;` : '')}
`;

const StyledSectionHeadButton = styled.div`
  padding: 0.375rem;
`;

const StyledSectionTitle = styled.div`
  padding: 0.75rem 0.375rem 0.75rem 0;
  flex-grow: 1;
`;

const StyledSectionIcon = styled.div`
  padding: 0.75rem 0;
  margin-right: 0.75rem;
`;

const StyledSectionContent = styled.div`
  font-size: var(--font-size-300);
  line-height: var(--line-height-300);
`;

const StyledSectionItem = styled.div``;

export enum MenuSectionVariant {
  default = 'default',
  minimal = 'minimal',
}

export interface SectionProps {
  items: React.ReactElement[];
  title?: string;
  icon?: React.ReactElement;
  headOptions?: {
    background?: string;
    color?: string;
    uppercase?: boolean;
  };
  variant?: MenuSectionVariant;
  button?: React.ReactElement;
}

export const MenuSection: React.FC<SectionProps> = ({
  title,
  icon,
  items,
  headOptions,
  variant = MenuSectionVariant.default,
  button,
}: SectionProps) => {
  return (
    <StyledSection variant={variant}>
      {title && (
        <StyledSectionHead {...headOptions}>
          {icon && <StyledSectionIcon>{icon}</StyledSectionIcon>}
          <StyledSectionTitle>{title}</StyledSectionTitle>
          {button && <StyledSectionHeadButton>{button}</StyledSectionHeadButton>}
        </StyledSectionHead>
      )}
      <StyledSectionContent>
        {items.map((item, index) => (
          <StyledSectionItem key={index}>{item}</StyledSectionItem>
        ))}
      </StyledSectionContent>
    </StyledSection>
  );
};
