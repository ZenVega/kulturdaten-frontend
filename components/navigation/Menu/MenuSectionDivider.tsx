import styled from '@emotion/styled';

export const MenuSectionDivider = styled.div`
  border-bottom: 1px solid var(--grey-400);
`;
