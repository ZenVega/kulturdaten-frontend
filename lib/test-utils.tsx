import { render, RenderOptions } from '@testing-library/react';
import { WindowContextProvider } from '../lib/WindowService';

const AllTheProviders: React.FC = ({ children }: { children: React.ReactNode }) => {
  return <WindowContextProvider>{children}</WindowContextProvider>;
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, 'queries'>) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
