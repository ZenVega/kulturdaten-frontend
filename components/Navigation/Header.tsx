import styled from '@emotion/styled';
import { useState } from 'react';
import { Breakpoint, useBreakpointOrWider } from '../../lib/WindowService';
import { MenuButton, MenuButtonState } from './MenuButton';

const HeaderContainer = styled.header`
  width: 100%;
  background: var(--grey-200);
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 0.75rem;
  border: 1px solid var(--grey-400);
  font-weight: var(--font-weight-bold);
  font-size: var(--font-size-300);
  line-height: var(--line-height-300);
`;

const HeaderLink = styled.a`
  text-decoration: none;
  color: inherit;
`;

interface HeaderProps {
  title: string;
  Link: React.FC<{ content: React.ReactElement }>;
  openState?: [boolean, React.Dispatch<React.SetStateAction<boolean>>];
}

export const Header: React.FC<HeaderProps> = ({ title, Link, openState }: HeaderProps) => {
  const initialState = useState<boolean>(false);
  const [menuOpen, setMenuOpen] = openState || initialState;

  const isWideOrWider = useBreakpointOrWider(Breakpoint.wide);

  return (
    <HeaderContainer>
      {!isWideOrWider ? (
        <MenuButton
          onClick={() => {
            setMenuOpen(!menuOpen);
          }}
          state={menuOpen ? MenuButtonState.close : MenuButtonState.open}
        />
      ) : (
        ''
      )}

      <Link content={<HeaderLink>{title}</HeaderLink>} />
    </HeaderContainer>
  );
};
