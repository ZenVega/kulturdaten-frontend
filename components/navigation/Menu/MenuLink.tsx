import Link from 'next/link';
import styled from '@emotion/styled';

import { useIsRouteStringActive } from '../../../lib/routing';
import { useContext } from 'react';
import { NavigationContext } from '../NavigationContext';
import { insetBorder } from '../../globals/Constants';
import { File } from 'react-feather';

const StyledA = styled.a<{ active?: boolean }>`
  color: inherit;
  text-decoration: ${({ active }) => (active ? 'underline' : 'none')};
  display: flex;
  align-items: center;
  width: 100%;
  padding: 0.75rem;
  box-shadow: ${insetBorder(true, false, false)};
  background: var(--white);

  &:hover {
    text-decoration: underline;
    background: var(--grey-350);
  }

  svg {
    display: inline-block;
    margin: 0 0.75rem 0 0;
    padding: 0;
    flex-shrink: 0;
    width: 1.5rem;
    height: 1.5rem;
    color: var(--black);
  }
`;

export enum MenuLinkType {
  internal = 'internal',
  external = 'external',
}

interface InternalMenuLinkProps {
  title: string;
  href: string;
  active?: boolean;
}

export interface MenuLinkProps extends InternalMenuLinkProps {
  type?: MenuLinkType;
}

const InternalMenuLink: React.FC<InternalMenuLinkProps> = ({
  title,
  href,
  active,
}: InternalMenuLinkProps) => {
  const isRouteActive = useIsRouteStringActive(href);
  const linkIsActive = active || isRouteActive;
  const { setNavigationOpen } = useContext(NavigationContext);

  return (
    <Link href={href} passHref>
      <StyledA title={title} onClick={() => setNavigationOpen(false)} active={linkIsActive}>
        <File />
        <span>{title}</span>
      </StyledA>
    </Link>
  );
};

export const MenuLink: React.FC<MenuLinkProps> = (props: MenuLinkProps) => {
  const { type = MenuLinkType.internal } = props;

  switch (type) {
    case MenuLinkType.internal: {
      return <InternalMenuLink {...props} />;
    }

    case MenuLinkType.external: {
      const { title, href } = props;
      return (
        <StyledA href={href} rel="noopener noreferrer" target="_blank">
          {title}
        </StyledA>
      );
    }

    default: {
      throw new Error(`MenuLink type "${type}" is not valid`);
    }
  }
};
