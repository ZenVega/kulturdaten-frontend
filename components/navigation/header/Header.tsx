import { css } from '@emotion/react';
import styled from '@emotion/styled';
import { useContext } from 'react';
import { Maximize2, Minimize2 } from 'react-feather';
import { useT } from '../../../lib/i18n';
import { Breakpoint, useBreakpointOrWider, WindowContext } from '../../../lib/WindowService';
import { Button, ButtonVariant, IconPosition } from '../../button';
import { mq } from '../../globals/Constants';
import { NavigationContext } from '../NavigationContext';

const StyledHeader = styled.header`
  width: 100%;
  background: var(--grey-200);
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  border-bottom: 1px solid var(--grey-400);

  ${mq(Breakpoint.mid)} {
    box-shadow: none;
  }
`;

const StyledLink = styled.a`
  font-weight: var(--font-weight-bold);
  font-size: var(--font-size-400);
  line-height: var(--line-height-400);
  text-decoration: none;
  color: inherit;
  padding: 0.75rem;

  ${mq(Breakpoint.mid)} {
    padding: 1.125rem 0.75rem;
  }
`;

const StyledHeaderButton = styled.div`
  padding: 0.375rem 0.75rem;

  ${mq(Breakpoint.mid)} {
    padding: 0.75rem;
  }
`;

const StyledExpandableButton = styled.button`
  appearance: none;
  border: none;
  margin: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background: var(--black);
  border-left: 1px solid var(--grey-400);
  align-self: stretch;
  width: 3.75rem;
  cursor: pointer;
  color: var(--white);

  &:hover {
    background: var(--grey-600);
  }
`;

interface HeaderProps {
  title: string;
  Link: React.FC<{ children: React.ReactElement<HTMLAnchorElement> }>;
  expandable?: boolean;
  subMenuKey?: string;
  defaultMenuKey: string;
  activeMenuTitle: string;
}

export const Header: React.FC<HeaderProps> = ({
  title,
  Link,
  expandable,
  subMenuKey,
  defaultMenuKey,
  activeMenuTitle,
}: HeaderProps) => {
  const {
    navigationOpen,
    setNavigationOpen,
    menuExpanded,
    setMenuExpanded,
    setActiveMenuKey,
    activeMenuKey,
  } = useContext(NavigationContext);
  const { rendered } = useContext(WindowContext);
  const isMidOrWider = useBreakpointOrWider(Breakpoint.mid);
  const t = useT();

  const isDefaultMenu = typeof activeMenuKey === 'undefined' || activeMenuKey === defaultMenuKey;

  const renderedLink = (
    <Link>
      <StyledLink>{title}</StyledLink>
    </Link>
  );

  const renderedButton =
    !isDefaultMenu || (!isMidOrWider && !navigationOpen) ? (
      <StyledHeaderButton>
        <Button
          variant={ButtonVariant.minimal}
          onClick={() => {
            setActiveMenuKey(defaultMenuKey);
            setNavigationOpen(true);
          }}
          icon="ChevronLeft"
          iconPosition={IconPosition.left}
          css={css`
            background: var(--white);
          `}
        >
          {t('menu.main')}
        </Button>
      </StyledHeaderButton>
    ) : undefined;

  const renderedSubMenuButton = (
    <StyledHeaderButton>
      <Button
        variant={ButtonVariant.minimal}
        onClick={() => {
          setActiveMenuKey(subMenuKey);
          setNavigationOpen(true);
        }}
        icon="ChevronLeft"
        iconPosition={IconPosition.left}
        css={css`
          background: var(--white);
        `}
      >
        {activeMenuTitle}
      </Button>
    </StyledHeaderButton>
  );

  return (
    <StyledHeader>
      {rendered && !isMidOrWider
        ? navigationOpen || typeof subMenuKey === 'undefined'
          ? renderedButton
          : renderedSubMenuButton
        : renderedButton}

      {rendered &&
        (isMidOrWider || !expandable) &&
        (!isMidOrWider || !renderedButton) &&
        renderedLink}

      {rendered && isMidOrWider && expandable && (
        <StyledExpandableButton onClick={() => setMenuExpanded(!menuExpanded)}>
          {menuExpanded ? <Minimize2 /> : <Maximize2 />}
        </StyledExpandableButton>
      )}
    </StyledHeader>
  );
};
