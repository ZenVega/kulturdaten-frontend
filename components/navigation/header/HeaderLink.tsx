import Link from 'next/link';

import { useLocale, routes } from '../../../lib/routing';
import { useOrganizerId } from '../../../lib/useOrganizer';

export interface HeaderLinkProps {
  children: React.ReactElement<HTMLAnchorElement>;
}

export const HeaderLink: React.FC<HeaderLinkProps> = ({
  children,
}: {
  children: React.ReactElement;
}) => {
  const locale = useLocale();
  const organizerId = useOrganizerId();

  return (
    <Link href={routes.dashboard({ locale, query: { organizer: organizerId } })} passHref>
      {children}
    </Link>
  );
};
