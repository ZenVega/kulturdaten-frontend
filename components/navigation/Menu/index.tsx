import styled from '@emotion/styled';
import React, { useContext } from 'react';
import { Button, ButtonSize, ButtonVariant, IconPosition } from '../../button';
import { NavigationContext } from '../NavigationContext';
import { MenuFolder } from './MenuFolder';
import { MenuIcon, MenuIconName } from './MenuIcon';
import { MenuLink, MenuLinkProps } from './MenuLink';
import { MenuSection, MenuSectionVariant } from './MenuSection';
import { MenuSectionDivider } from './MenuSectionDivider';

export enum MenuItem {
  link = 'link',
  folder = 'folder',
  button = 'button',
  divider = 'divider',
}

export type MenuItemLink = {
  title: string;
  href: string;
  active?: boolean;
};

export type MenuItemButton = {
  label: string;
  onClick: () => void;
  icon?: string;
  iconPosition?: IconPosition;
};

export type MenuItemFolder = {
  label: string;
  menuKey: string;
};

export type MenuData = {
  key: string;
  expandable: boolean;
  title: string;
  List?: React.FC<{ expanded?: boolean }>;
  sections?: MenuSectionData[];
};

export type MenuSectionData = {
  title: string;
  items: {
    type: MenuItem;
    action?: MenuItemLink | MenuItemButton | MenuItemFolder;
  }[];
  button?: React.ReactElement;
  icon?: MenuIconName;
  headOptions?: {
    background?: string;
    color?: string;
    uppercase?: boolean;
  };
  variant?: MenuSectionVariant;
};

const StyledMenuSections = styled.div`
  padding: 1.5rem 0.75rem;
  display: grid;
  grid-template-columns: 100%;
  grid-row-gap: 1.5rem;
`;

const StyledMenu = styled.div``;

interface MenuProps {
  menuData: MenuData;
}

export const Menu: React.FC<MenuProps> = ({ menuData }: MenuProps) => {
  const { setNavigationOpen, menuExpanded } = useContext(NavigationContext);

  const { sections, List } = menuData;

  const renderedSections = sections?.map(
    ({ title, icon, items, headOptions, variant, button }, index) => {
      const renderedItems = items?.map(({ type, action }, actionIndex) => {
        switch (type) {
          case MenuItem.link: {
            return <MenuLink key={actionIndex} {...(action as MenuLinkProps)} />;
          }

          case MenuItem.button: {
            const { label, onClick, icon, iconPosition } = action as MenuItemButton;
            return (
              <Button
                onClick={() => {
                  onClick();
                  setNavigationOpen(false);
                }}
                variant={ButtonVariant.minimal}
                size={ButtonSize.small}
                icon={icon}
                iconPosition={iconPosition}
              >
                {label}
              </Button>
            );
          }

          case MenuItem.folder: {
            const { label, menuKey } = action as MenuItemFolder;
            return <MenuFolder label={label} menuKey={menuKey} />;
          }

          case MenuItem.divider: {
            return <MenuSectionDivider />;
          }

          default: {
            break;
          }
        }
      });

      return (
        <MenuSection
          title={title}
          icon={icon ? <MenuIcon type={icon} /> : undefined}
          items={renderedItems}
          key={index}
          headOptions={headOptions}
          variant={variant}
          button={button}
        />
      );
    }
  );

  return (
    <StyledMenu>
      {List && React.createElement(List, { expanded: menuExpanded })}
      {renderedSections && <StyledMenuSections>{renderedSections}</StyledMenuSections>}
    </StyledMenu>
  );
};
