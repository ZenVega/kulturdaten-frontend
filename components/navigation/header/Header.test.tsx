import { render } from '../../../lib/test-utils';
import { MenuItemType } from '../Menu';
import { HeaderMain } from './Header';
import { HeaderLink } from './HeaderLink';

test('Header renders', () => {
  const { container } = render(
    <HeaderMain
      title="Test title"
      Link={HeaderLink}
      menuItems={[
        { type: MenuItemType.link, action: { href: '#', title: 'Test' } },
        { type: MenuItemType.divider },
      ]}
      user={undefined}
      userIsLoggedIn={false}
    />
  );
  expect(container).toMatchSnapshot();
});
