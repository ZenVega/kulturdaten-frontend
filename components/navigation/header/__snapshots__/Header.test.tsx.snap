// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders 1`] = `
<div>
  <header
    class="css-1p8v3tm-HeaderContainer e1sn27871"
  >
    <button
      aria-label="Menü öffnen"
      class="css-1290inr-StyledMenuButton e91bmyz0"
      title="Menü öffnen"
    >
      <svg
        aria-labelledby="menu-button-icon"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title
          id="menu-button-icon"
        >
          Menü öffnen
        </title>
        <g>
          <path
            d="M5 8H19"
            stroke="#111111"
            stroke-linecap="round"
            stroke-width="2"
          />
        </g>
        <g>
          <path
            d="M19 16H5"
            stroke="#111111"
            stroke-linecap="round"
            stroke-width="2"
          />
        </g>
      </svg>
    </button>
    <a
      class="css-1kagrgy-StyledLink e1sn27870"
      href="/undefined"
    >
      Test title
    </a>
  </header>
   
</div>
`;
